compiler-rt/lib/mpk_untrusted/mpk_formatter.cpp

```diff
@@ -1,42 +1,116 @@
 #include "mpk_formatter.h"
+#include "alloc_site_handler.h"
+#include <bits/stdint-intn.h>
+#include <bits/stdint-uintn.h>
+#include <fstream>
 
 namespace __mpk_untrusted {
 
-// Flush Allocs is to be called on program exit to flush all faulting
-// allocations to disk/file.
-void flushAllocs() {
-  auto handler = AllocSiteHandler::init();
+#define ATTEMPTS 128
 
-  std::vector<llvm::json::Value> jAllocVec;
-  for (auto alloc : handler->faultingAllocs()) {
-    jAllocVec.push_back(toJSON(alloc));
+bool makeUniqueFile(std::string Model, std::ofstream &OS) {
+  srand(time(NULL));
+  // Loop for number of attempts just in case level of entropy is to low
+  for (uint8_t attempt = 0; attempt != ATTEMPTS; ++attempt) {
+    std::string unique = Model;
+    for (uint32_t i = 0, e = unique.length(); i != e; ++i) {
+      if (unique[i] == '%')
+        unique[i] = "0123456789abcdef"[rand() & 15];
+    }
+    struct stat info;
+    // If file does not already exist, create and return ofstream
+    if (stat(unique.c_str(), &info) == -1) {
+      OS.open(unique, std::ios_base::trunc);
+      // Ensure we correctly opened OS ofstream
+      if (OS)
+        return true;
+    }
   }
 
-  if (jAllocVec.empty()) {
-    // TODO : This is for testing only to make sure JSON output works.
-    __sanitizer::Report("INFO : Faulting alloc vector is empty, so we are "
-                        "making a fault entry for testing.\n");
-    jAllocVec.push_back(toJSON(AllocSite::error()));
+  // Failed to make unique name
+  return false;
+}
 
-    // __sanitizer::Report("INFO : No faulting instructions to export,
-    // returning.\n"); return;
+bool is_directory(std::string directory) {
+  struct stat info;
+  if (stat(directory.c_str(), &info) != 0)
+    return false;
+  else if (info.st_mode & S_IFDIR)
+    return true;
+  else
+    return false;
+}
+
+// Function for handwriting the JSON output we want (to remove dependency on
+// llvm/Support).
+void writeJSON(std::ofstream &OS, std::set<AllocSite> &faultSet) {
+  OS << "[\n";
+  int64_t items_remaining = faultSet.size();
+  for (auto fault : faultSet) {
+    --items_remaining;
+    if (items_remaining <= 0) {
```

Comment:
can you rewrite this so that only the last comma and newline are in the conditional. it will be easier to audit later, and I actually had a hard time finding the difference in the branches since they only differ by one character.

ternary operator might be your friend here. i.e ` << last_item ? "" : "," << `

Reply:
if using ternary operator, you might need to store it in a string first, which is unfortunate.

Reply:
Switched to using ternary operator as suggested.